flow/directives.py

```diff
@@ -0,0 +1,315 @@
+        return value
+    nranks = directives.get('nranks', 1)
+    omp_num_threads = directives.get('omp_num_threads', 1)
+    if callable(nranks) or callable(omp_num_threads):
+        return value
+    else:
+        return max(value, max(1, nranks) * max(1, omp_num_threads))
+
+
+natural_number = OnlyType(int, postprocess=raise_below(1))
+# Common directives and their instantiation as _DirectivesItem
+NP = _DirectivesItem('np', natural_number, _NP_DEFAULT, finalize=finalize_np)
+NP.__doc__ = """
+The number of tasks expected to run for a given operation
+
+Expects a natural number (i.e. an integer >= 1). When querying
+this directive, introspection into the Directives is used to determine if the
+product of "nranks" or "omp_num_threads" is greater than the current value. The
+maximum of these two values is used.
+"""
+
+nonnegative_int = OnlyType(int, postprocess=raise_below(0))
+NGPU = _DirectivesItem('ngpu', nonnegative_int, 0)
+NGPU.__doc__ = """
+The number of GPU's to request for this operation.
+
+Expects a nonnegative integer.
+"""
+NRANKS = _DirectivesItem('nranks', nonnegative_int, 0)
+NRANKS.__doc__ = """
+The number of MPI ranks to use for this operation.
+
+Expects a nonnegative integer.
+"""
+
+OMP_NUM_THREADS = _DirectivesItem('omp_num_threads', nonnegative_int, 0)
+OMP_NUM_THREADS.__doc__ = """
+The number of OpenMP threads to use for this operation.
+
+Expects a nonnegative integer.
+"""
+
+
+def no_aggregation(v, o):
+    return v
+
+
+EXECUTABLE = _DirectivesItem('executable', OnlyType(str), sys.executable,
+                             no_aggregation, no_aggregation)
+EXECUTABLE.__doc__ = """
+The location in the file system to the executable to be used for this operation.
+
+Expects a string pointing to a valid executable file in the
+current file system.
+
+By default this should point to a Python executable (interpreter); however, if
+the :py:class:`FlowProject` path is set to an empty string, the executable can
+be the file path to an executable Python script.
+"""
+
+
+nonnegative_real = OnlyType(float, postprocess=raise_below(0))
+WALLTIME = _DirectivesItem('walltime', nonnegative_real, 12.,
+                           lambda x, y: x + y, max)
+WALLTIME.__doc__ = """
+The number of hours to request for executing this job.
+
+This directive expects a float representing the number hours including
+fraction's of hours.
+"""
+
+MEMORY = _DirectivesItem('memory', natural_number, 4)
+MEMORY.__doc__ = """
+The number of GB to request for an operation.
+
+Expects a natural number (i.e. an integer >= 1).
+"""
+
+
+def is_fraction(value):
+    if 0 <= value <= 1:
+        return value
+    else:
+        raise ValueError("Value must be beween 0 and 1.")
+
+
+PROCESS_FRACTION = _DirectivesItem('processor_fraction',
+                                   OnlyType(float, postprocess=is_fraction),
+                                   1., no_aggregation, no_aggregation)
+PROCESS_FRACTION.__doc__ = """
+Needs to be filled out.
```

Comment:
This looks like a TODO item. I don't know what this directive does...

Reply:
I don't either honestly. I have never used it, but it is already in the code base, so I added the directive as well as I could.

Reply:
This directive exists to improve launcher support on stampede. Specifically, this variable controls how much cpu needs to be allocated to a single job op when using stampede. For instance, if the value is 0.5 and you bundle 20, it will request 10 cpus.

Reply:
@vyasr thanks. In that case, if this is only for Stampede we should move this to only be in the Stampede environment.

Reply:
I don't think that's necessary. The launcher on stampede is basically identical to a process pool (e.g. `multiprocessing.Pool` in python) except that it transparently supports things like multi-node with MPI on stampede. I'm not sure how well Python handles such use cases (although it might be fine if you bind all cores to one task), but since Stampede2 provides a specific solution it made sense to take advantage.  The functionality is general, though (I believe it's just used as a prefactor in the resource calculation template filter), and could conceivably be used by someone running a process pool directly with python, for instance, where they anticipate being able to run a large number of job-ops through a small number of cores if each job-op is small. So I think it makes sense to leave it here, but simply add documentation indicating this.